### ISSUE TITLE
icc_transform + heifsave results in segmentation fault, only when using bundled libvips

### ISSUE THREAD

**Issue by premist** [akash-akya/vix#109]
Hello, I'm experiencing a strange error.

When I simply load an image and save it as an avif format using `Vix.Vips.Operation.heifsave/3`, it works without issues. 
However, when I use `Vix.Vips.Operation.icc_transform/3` to convert image and use `heifsave/3`, it doesn't work. `jpegsave/3` works though, which makes it even stranger.  

This issue goes away when using `VIPS_COMPLIATION_MODE=PLATFORM_PROVIDED_LIBVIPS`, so that is a workaround I'm using at this moment.

### Minimal repro code

```elixir
{:ok, img} = Vix.Vips.Image.new_from_file("test.jpg")
{:ok, newimg} = Vix.Vips.Operation.icc_transform(img, "DisplayP3-v4.icc", embedded: true, depth: 16)
Vix.Vips.Operation.heifsave(newimg, "vix-modified.avif", lossless: true, compression: :VIPS_FOREIGN_HEIF_COMPRESSION_AV1)
```

Output
```bash 
❯ elixir minimal.exs
zsh: segmentation fault  elixir test.exs
```

### GitHub Actions Tests

I have tested the above using GitHub Actions, using Linux and macOS, with bundled libvips and system libvips.

- Sample code: https://github.com/premist/vix-test/blob/main/test.exs
- GitHub Actions workflow: https://github.com/premist/vix-test/blob/main/.github/workflows/run.yml
- GitHub Actions run results for both Linux and macOS: https://github.com/premist/vix-test/actions/runs/4765898786

<img width="339" alt="image" src="https://user-images.githubusercontent.com/291078/233666577-80445cf3-1655-49ac-83df-1e7433b43000.png">

**Comment by akash-akya** [comment]
Hey @premist, thanks for the details issue report with test cases, it makes my job simpler.

I think the issue is because pre-compiled libvips does not support writing higher bit depth HEIF images. It is disabled during compilation [via flag](https://github.com/lovell/sharp-libvips/blob/main/build/lin.sh#L258). This is nothing to do with `icc_transform`. 

Following test fail too with segfault

```elixir
alias Vix.Vips.Operation
alias Vix.Vips.Image

# create 16 bitdepth image for testing
{:ok, img} = Image.new_from_file("test.jpg")

:ok =
  img
  |> Operation.icc_transform!("DisplayP3-v4.icc", depth: 16)
  |> Operation.pngsave!("vix-modified.png")

# try to write 16bit depth image 
{:ok, img} = Image.new_from_file("vix-modified.png")
:ok = Operation.heifsave!(img, "vix-modified.avif")
```

But if you comment, `  |> Operation.icc_transform!("DisplayP3-v4.icc", depth: 16)` it would succeed. 

Either way we should not see segfault even if the higher bit depth is not supported. I'll raise this to upstream and try to find a solution. Meanwhile, please continue using platform provided libvips

**Comment by akash-akya** [comment]
@premist It looks like support for higher bit depth is disabled to improve the binary size.
I built a new libvips pre-built package, can you help to test?

```
{:vix, github: "akash-akya/vix", branch: "dev"}
```

**Comment by ringofhealth** [comment]
I am also running into this issue, however it does not seem to be caused by that particular transform. I did notice that when I upgraded to 0.18 while vips is is on 8.13.3, there is alot more seg faults. when I downgraded back to 0.14. its more stable. not sure this helps

**Comment by akash-akya** [comment]
@ringofhealth is it possible to know what were the operations? (what format you trying to write, etc.) Issue might be something similar to this, pre-built libvips does not come with all the operations and functionality which the platform provided libvips comes with.

**Comment by ringofhealth** [comment]
yep, here is the conversion module

```
defmodule Service.VipConverter do
  alias Vix.Vips.Image
  alias Vix.Vips.Operation
  require Logger
  @moduledoc false

  def convert(input, output, {:blur, size, blur_sigma, quality}) do
    # blur_sigma has to be a float
    with {:ok, vipsimg} <- Image.new_from_file(input),
         max_width when not is_nil(max_width) <- Image.width(vipsimg),
         {:ok, thumb} <- Operation.resize(vipsimg, size / max_width),
         {:ok, final} <- Operation.gaussblur(thumb, blur_sigma),
         :ok <- Vix.Vips.Image.write_to_file(final, output <> "[Q=#{quality},strip]") do
      {:ok, output}
    else
      {:error, _reason} = e ->
        Logger.error("#{__MODULE__}: #{inspect(e)}")

      e ->
        Logger.error("#{__MODULE__}: #{inspect(e)}")

        {:error, e}
    end
  end

  def convert(input, output, {:resize, size, quality}) do
    # blur_sigma has to be a float
    with {:ok, vipsimg} <- Image.new_from_file(input),
         max_width when not is_nil(max_width) <- Image.width(vipsimg),
         {:ok, thumb} <- Operation.resize(vipsimg, size / max_width),
         :ok <- Vix.Vips.Image.write_to_file(thumb, output <> "[#{quality},strip]") do
      {:ok, output}
    else
      {:error, _reason} = e ->
        Logger.error("#{__MODULE__}: #{inspect(e)}")

        e

      e ->
        Logger.error("#{__MODULE__}: #{inspect(e)}")

        {:error, e}
    end
  end

  def convert(input, output, {:resize_watermark, size, user, quality}) do
    watermark = ~s(<span foreground="#ffffffd9">#{root_domain() <> user}</span>)

    with {:ok, vipsimg} <- Image.new_from_file(input),
         max_width when not is_nil(max_width) <- Image.width(vipsimg),
         max_height when not is_nil(max_height) <- Image.height(vipsimg),
         {:ok, {original_text, _}} <-
           Operation.text(
             watermark,
             fontfile: font_file(),
             font: "Figtree Medium",
             width: calculate_font_width(max_width, max_height),
             rgba: true,
             # have large max height to let libvips auto set it
             height: max_height,
             # to avoid line splitting
             spacing: max_height
           ),
         {:ok, overlay} <-
           Operation.composite2(
             vipsimg,
             original_text,
             :VIPS_BLEND_MODE_OVER,
             x: round(Image.width(vipsimg) / 2 - Image.width(original_text) / 2),
             y: round(max_height - max_height / 10)
           ),
         {:ok, final} <-
           Operation.resize(overlay, size / max_width),
         :ok <- Vix.Vips.Image.write_to_file(final, output <> "[#{quality},strip]") do
      {:ok, output}
    else
      {:error, _reason} = e ->
        Logger.error("#{__MODULE__}: #{inspect(e)}")

        e

      e ->
        Logger.error("#{__MODULE__}: #{inspect(e)}")

        {:error, e}
    end
  end

  defp calculate_font_width(max_w, max_h) do
    if max_w >= max_h do
      round(max_h * 0.4)
    else
      round(max_w * 0.5)
    end
  end

  def font_file() do
    Application.app_dir(:helio, "priv/fonts/Figtree-Medium.ttf")
  end
end

```

**Comment by premist** [comment]
@akash-akya tested with dev branch, works well on all platforms! https://github.com/premist/vix-test/actions/runs/4787714807

**Comment by akash-akya** [comment]
Thanks @premist for testing and confirming. I'll cut a release soon with the newer binaries. It will consume a bit more disk-space, but I think that's fine.

@ringofhealth thanks for the code. Is it possible to share sample cases? like function parameters when it crashes. 
Btw if the pre-built binary is causing issues, you can still use latest vix with platform provided libvips. Just set `VIX_COMPILATION_MODE` env variable to `PLATFORM_PROVIDED_LIBVIPS`. No need to still use old vix version.

**Comment by ringofhealth** [comment]
Ill dig up a coredump when the next time it crashes. My top suspect is that there might be unsafe memory usage by vips. I am doing parallel processing of image upload with vips. Its a concurrent worker load of 30 workers. If I go more, I see more crashes, it seems to be relatively stable under 30 works concurrently. 

is the prebuild binary have webp support? as well as pango2? I have some watermarking and overlaying work I do

**Comment by akash-akya** [comment]
> is the prebuild binary have webp support? as well as pango2? I have some watermarking and overlaying work I do

Yes. Prebuilt binary support both webp & pango

> Ill dig up a coredump when the next time it crashes. My top suspect is that there might be unsafe memory usage by vips. I am doing parallel processing of image upload with vips. Its a concurrent worker load of 30 workers. If I go more, I see more crashes, it seems to be relatively stable under 30 works concurrently.

I see, it might also be that we are hitting some limit and that is not getting handled properly and causing issue (number open FD, memory, some timeout etc.)
By `30 workers` you mean beam schedulers or `VIPS_CONCURRENCY`? If possible, can you share mem-usage/limit, cup-usage/limit and other resource usage info?

**Comment by ringofhealth** [comment]
> > is the prebuild binary have webp support? as well as pango2? I have some watermarking and overlaying work I do
> 
> Yes. Prebuilt binary support both webp & pango
> 
> > Ill dig up a coredump when the next time it crashes. My top suspect is that there might be unsafe memory usage by vips. I am doing parallel processing of image upload with vips. Its a concurrent worker load of 30 workers. If I go more, I see more crashes, it seems to be relatively stable under 30 works concurrently.
> 
> I see, it might also be that we are hitting some limit and that is not getting handled properly and causing issue (number open FD, memory usage, timeout etc.) By `30 workers` you mean beam schedulers or `VIPS_CONCURRENCY`? If possible, can you share memory usage/limit, CPU usage/limit and other resource usage info?

yeah let me clarify, by 30 works, I mean `oban` workers, each of the worker just runs a function for processing image. for each 1 image/video, this function generates 6 derivatives versions(sizes, watermarks) 

so If I was doing image processing with vips for 30 workers,

lets take the `convert` function in that module above I pasted, there is 6 VIPS operation in there

Its could be doing as much 30 workers * 6 derivative * 6 vips = 1080 Vips operations lol

**Comment by akash-akya** [comment]
I see, I wonder if any of the system limits getting hit. I remember seeing segfaults when the number of open FD limit was reached while testing my other library. 
I will try to reproduce this as soon as I find some time.

**Comment by akash-akya** [comment]
I'll look into the crash on high load issue separately #111 

Current issue is fixed by #110
